generator/src/main/java/com/graphicsfuzz/generator/fuzzer/OpaqueExpressionGenerator.java

```diff
@@ -1418,6 +1421,91 @@ public boolean preconditionHolds(Expr expr, BasicType basicType, boolean constCo
     }
   }
 
+  /**
+   * Identity transformation to insert a vector or matrix into a larger vector or matrix using a
+   * type constructor, then extract it back out again using another type constructor. When
+   * performed, transforms an expression e with type t -> identity(t(identity(m(identity(e), ..)))),
+   * where m is a type of equal or greater width than t.
+   * The rules for this smaller -> larger -> smaller transformation are as follows:
+   *     If the given type is a vector, the inner constructor type can be a vector of the same
+   *     element type with equal/greater width than the given type, or (provided that the given type
+   *     is a floating point genType) it can be a matrix with equal/greater column width than the
+   *     given type.
+   *     If the given type is a matrix, the inner constructor type can be a matrix with
+   *     equal/greater column/row width than the given type.
+   */
+  private class IdentityCompositeConstructorExpansion extends AbstractIdentityTransformation {
```

Comment:
Below, and in the above comment, would there be any harm in applying this to scalars too - i.e. pushing a scalar into a bigger type and then pulling it out again?

E.g.:

  float v = float(mat2x4(1.0));

is valid and should give 1.0.

Reply:
Yeah, you're right! Additionally, I barely even have to change the logic to accommodate scalars.